### DISCUSSION TITLE
PostFX Later Category definitions

### DISCUSSION THREAD

**Discussion by HogJonny-AMZN** [o3de/o3de#4101] [General]
O3DE provides an ordered stack for PostFX Layer Categories.  PostFX components (like Bloom, Dof, etc.) can be masked via mechanisms like volumes/falloff/gradients, so look modifications can be blended.  There are two aspects of ordering, the first is 'category' (higher order bucket sorting), the second is 'priority (ordering within each category bucket).

This determines the order in which PostFX Layer data is used to used to resolve the deterministic blending of params for various PostFX components.

The default categories and their ordering is set in the following file:
"C:\Depot\o3de-engine\Gems\AtomLyIntegration\CommonFeatures\Assets\PostProcess\default.postfxlayercategories"

The current categories are:
1. Default: (implied priority, max 2,147,483,647)
2. Level: 1000
3. Camera: 100
4. Volume: 10

The explain this, here is what is happening:
- the default layer is always active ... think of it like the canvas, everything else is on top.  If we are thinking along the lines of the Photoshop Layer stack, this would be the initial 'background layer'.
- All PostFX components when initially added, are assigned to the default category.
- The priority always defaults to 1
- If more then one PostFX component of the same type (like Bloom) are both active and are assigned the same category, and have the same priority, this may break being deterministic (the priority provides a mechanism by which user can ensure determinism.)
- Currently, something assigned to the Level category, will override the default layer, etc.

Users can add .postfxlayercategories file to their project, so they can extend the number and ordering of pre-defined categories.  And the crux here is that with the existing numerical ordering there is not much room above volumes (priority 10) to add new categories (just to point out opne flaw)

The intent with this discussion, is to update this an provide better names and default ordering, which are more inline with the state of a game.
1. Default:          A default global look
2. Level:             N# of World design and env based looks (like forest, mountain peak, etc.)
3. Volume:         Location based looks (like a village, or a particular monument)
3. Camera:         A look for a specific camera / view
4. Gameplay:     Gameplay driven / dynamic event looks (like a grenade flash)
5. Cinematics:    Sequence / Cinematic keyframe based looks (ease into cinematic post, fade back out to previous looks on exit)
6. Front End:      Front End UI Looks, such as pause screen, player lobby, etc.

The gist is, that we can have an open world game and the PostFX looks are dynamically driven in complex blended ways based on the player experience.  World looks as an example could be tied to a 'placement mask' (gradient signal) that shares data with the world (terrain, dynamic vegetation, etc.) The desert could be graded in a sepia tone, the forest could be punchy and slightly saturated, climbing in altitude could adjust color temp / white balance, etc.  But then a prefab, like the witches hut, could have it's own look blend/overrides, bound to a volume (shape) with falloff.  When a grenade goes off, it can lerp up/down it's look.  If the player triggers a cinematic, we can blend into then out of a keyframed look. And if the game is paused, we can desaturate and blur the look behind the front-end UI.

So the updated data will look like this:

1,000,000 - FrontEnd
2,000,000 - Cinematics
3,000,000 - Gameplay
4,000,000 - Camera
5,000,000 - Volume
6,000,000 - Level
2,147,483,647 - Default

We kept all existing category names, to not break any existing components/prefabs.

This better mimics the basic statefullness of the game runtime, which provides a more intuitive set of categories, which are more likely to be correctly ordered based on how they are activated, and significantly expands the space for padding (so users can better define their own sub-categories above or below any of the defaults)

**Comment by HogJonny-AMZN** [comment]
![image](https://user-images.githubusercontent.com/23222931/133325733-8c136127-292c-43f3-8936-44c2a085f887.png)

**Comment by HogJonny-AMZN** [comment]
PR: https://github.com/o3de/o3de/pull/4125

**Reply by xychen5** [reply]
hi，may i ask u a question? i wonder if postFX layer can achieve such a effect: a model with bloom effect, and another don't
